squids/message-indexer/src/subsquid/processor.ts

```diff
@@ -0,0 +1,102 @@
+import { plainToInstance } from 'class-transformer';
+import { BatchContext, BatchProcessorItem } from '@subsquid/substrate-processor';
+import { Store } from '@subsquid/typeorm-store';
+
+import { processor } from './init-processor';
+import { initDB } from './init-db';
+import { MessageType, SubSquidMessageEventType } from '../common/enum';
+import { Message } from '../model';
+
+const messageMap = new Map<string, Message>();
+
+async function eventHandler(ctx: BatchContext<Store, BatchProcessorItem<typeof processor>>): Promise<void> {
+  for (const block of ctx.blocks) {
+    for (const item of block.items) {
+      // eslint-disable-next-line no-continue
+      if (item.kind !== 'event') continue;
+
+      const { name } = item.event;
+      const { header: { timestamp, hash } } = block;
+      const time = new Date(timestamp);
+
+      if (name === SubSquidMessageEventType.GEAR_USER_MESS_READ) {
+        const { id, reason } = item.event.args;
+        const message = await ctx.store.findOneBy(Message, { id });
```

Comment:
It's better to use a temporary state to store data from each batch and save it into database only once for each batch. It will reduce number of r/w operations to the database as well as increase speed of network indexing.

Reply:
Premature optimization may not good idea sometimes. Currently in the `processor` for the `write option` in a database for each batch elements going only once. I didn't see any points right now to optimize if we don't have any information for that case.

Reply:
Of course it appears. I described the problem in the next sentence

> Because in most cases MessageQueued and MessagesDispatched events occur in one block so message that you're trying to get from database in line 40 won't exist.